libvast/src/system/archive.cpp

```diff
@@ -77,94 +77,124 @@ archive(archive_type::stateful_pointer<archive_state> self, path dir,
   });
   self->set_down_handler([=](const down_msg& msg) {
     VAST_DEBUG(self, "received DOWN from", msg.source);
-    self->state.active_exporters.erase(msg.source);
+    auto& st = self->state;
+    st.active_exporters.erase(msg.source);
+    while (st.unhandled_ids.front().second == msg.source)
+      st.unhandled_ids.pop();
   });
   if (auto a = self->system().registry().get(accountant_atom::value)) {
     namespace defs = defaults::system;
     self->state.accountant = actor_cast<accountant_type>(a);
     self->send(self->state.accountant, announce_atom::value, self->name());
     self->delayed_send(self, defs::telemetry_rate, telemetry_atom::value);
   }
-  return {[=](const ids& xs) -> caf::result<done_atom, caf::error> {
-            VAST_ASSERT(rank(xs) > 0);
-            VAST_DEBUG(self, "got query for", rank(xs),
-                       "events in range [" << select(xs, 1) << ','
-                                           << (select(xs, -1) + 1) << ')');
-            if (self->state.active_exporters.count(
-                  self->current_sender()->address())
-                == 0) {
-              VAST_DEBUG(self, "dismisses query for inactive sender");
-              return make_error(ec::no_error);
+  return {
+    [=](const ids& xs) {
+      VAST_ASSERT(rank(xs) > 0);
+      VAST_DEBUG(self, "got query for", rank(xs),
+                 "events in range [" << select(xs, 1) << ','
+                                     << (select(xs, -1) + 1) << ')');
+      auto requester = caf::actor_cast<receiver_type>(self->current_sender());
```

Comment:
I'm not sure if this is a safe operation. You cannot guarantee that the sender is `receiver_type`.

Reply:
We talked about this in person: `receiver_type` defines a behavior that the caller must necessarily implement. I have added a check to see whether the cast succeeded.